onnxruntime/core/providers/cpu/controlflow/if.cc

```diff
@@ -315,22 +336,34 @@ Status IfImpl::AllocateOutputTensors() {
         // we still need a value to put in the feeds we give to the execution frame, so just use an empty MLValue
         outputs_.push_back({AllocationType::Delayed, {}});
       }
-    } else if (graph_output_type->has_sequence_type() || is_optional_tensor_sequence) {
+    } else if (graph_output_type->has_sequence_type()
+#if !defined(DISABLE_OPTIONAL_TYPE)
+               || is_optional_tensor_sequence
+#endif
+    ) {
       auto* seq_tensor = context_.Output<TensorSeq>(index);
       if (!seq_tensor)
         return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Failed to create output tensor for ", graph_output->Name());
       outputs_.push_back({AllocationType::IfOutput, *context_.GetOutputMLValue(index)});
     } else {
-      // Shouldn't hit this
-      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Only tensors or sequence of tensors are supported");
+      // Shouldn't hit this as the kernel assignment logic should check for the types before assigning this kernel
+#if !defined(DISABLE_OPTIONAL_TYPE)
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "Only tensors, tensor sequence, optional tensor, and optional tensor sequence types are supported");
+#else
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "Only tensors and tensor sequence types are supported");
+#endif
     }
```

Comment:
Probably fine to leave the error message as is to keep it simple. If someone attempts to use a model with 'optional' support disabled the model won't load.